job/server/src/main/java/alluxio/job/plan/transform/CompactDefinition.java

```diff
@@ -80,24 +86,52 @@ private boolean shouldIgnore(URIStatus status) {
     AlluxioURI outputDir = new AlluxioURI(config.getOutput());
 
     List<URIStatus> files = Lists.newArrayList();
+    // use double to prevent overflow
+    double totalFileSize = 0;
     for (URIStatus status : context.getFileSystem().listStatus(inputDir)) {
       if (!shouldIgnore(status)) {
         files.add(status);
+        totalFileSize += status.getLength();
       }
     }
+
     Map<WorkerInfo, ArrayList<CompactTask>> assignments = Maps.newHashMap();
-    int groupSize = Math.max(1, (files.size() + 1) / config.getNumFiles());
+    int maxNumFiles = config.getMaxNumFiles();
+    long groupMinSize = config.getMinFileSize();
+    if (totalFileSize / groupMinSize > maxNumFiles) {
+      groupMinSize = Math.round(totalFileSize / maxNumFiles);
+    }
+
+    if (!files.isEmpty() && config.getPartitionInfo() != null) {
+      // adjust the group minimum size for source compression ratio
+      groupMinSize *= COMPRESSION_RATIO.get(
```

Comment:
what happens if the format is unknown? Maybe for unknown formats, it should just default to 1.0 or 1.5 or something.

Reply:
if the format is unknown, we have bigger problems.. we wouldn't even be able to read the file, right?

Reply:
in fact if the format is unknown, we throw an exception right now. which seems like the right thing to do

Reply:
I see. Does the `getFormat` already handle the cases where the files don't have nice file extensions?